File: server/demoDevice.js
// Smart Learning Center - demo-mode device generator.
// Mimics the ESP32's wire protocol (same 'connected' / 'disconnected' /
// 'message' / 'raw' events and send(line) method as bridge.js) so the rest
// of server.js does not need to know whether it is talking to the real
// Wokwi simulation or this generator. Used when BRIDGE_MODE=demo, so the
// dashboard can be rehearsed without Wokwi running. The UI must show a
// persistent DEMO MODE banner whenever this is active - see server.js.
import { EventEmitter } from 'events';

const ROOM = 'SF-03';

export default class DemoDevice extends EventEmitter {
  constructor() {
    super();
    this.connected = false;
    this.simMinutes = 7 * 60 + 10;
    this.scale = 1;
    this.registry = [];
    this.schedule = [];
    this.timer = null;

    this.state = {
      temp: 24, hum: 55, motion: false,
      smoke: 6, smokeLevel: 0,
      wastePct: 15, wasteFull: false,
      metalThreshold: 55, holdActive: false,
      doorsUnlocked: 0, overrideActive: false,
      emergencyState: 'IDLE',
      smartRoomState: 'STANDBY',
      attendanceToday: 0,
      wifiConnected: true, rssi: -58, ip: '10.0.0.42',
    };

    this.modules = {
      rfid: true, smart_room: true, environment: true, security: true,
      metal_detector: true, waste: true, network: true,
    };
  }

  start() {
    this.connected = true;
    setTimeout(() => this.emit('connected'), 50);
    this._emitMsg({ t: 'boot', device: 'DEMO-ESP32', fw: 'demo-1.0.0', modules: 'A,B,C,D,E,F,G,H' });
    this._emitMsg({ t: 'log', msg: 'Demo device generator started (BRIDGE_MODE=demo). No Wokwi simulation is running.' });
    this.timer = setInterval(() => this._tick(), 1000);
  }

  stop() {
    this.connected = false;
    if (this.timer) clearInterval(this.timer);
    this.emit('disconnected');
  }

  isConnected() { return this.connected; }

  send(line) {
    if (!line.startsWith('CMD ')) return false;
    let msg;
    try { msg = JSON.parse(line.slice(4)); } catch { return false; }
    this._handleCommand(msg);
    return true;
  }

  _emitMsg(obj) { this.emit('message', obj); }
  _flow(m, s, v) { this._emitMsg({ t: 'flow', m, s, v: v || undefined }); }

  _hhmm() {
    const h = Math.floor(this.simMinutes / 60) % 24;
    const m = this.simMinutes % 60;
    return String(h).padStart(2, '0') + ':' + String(m).padStart(2, '0');
  }

  _tick() {
    this.simMinutes = (this.simMinutes + this.scale) % 1440;

    const m = this.modules;

    if (m.environment) {
      this.state.temp += (Math.random() - 0.5) * 0.3;
      this.state.temp = Math.max(18, Math.min(32, this.state.temp));
      this.state.hum += (Math.random() - 0.5) * 0.6;
      this.state.hum = Math.max(30, Math.min(85, this.state.hum));
      this.state.motion = Math.random() < 0.3;
    }
    if (m.security) {
      this.state.smoke = Math.max(2, Math.min(18, this.state.smoke + (Math.random() - 0.5) * 2));
    }
    if (m.waste && !this.state.wasteFull) {
      this.state.wastePct = Math.min(100, this.state.wastePct + 0.04);
      this.state.wasteFull = this.state.wastePct >= 80;
    }

    const active = m.smart_room ? this.schedule.find((s) => this.simMinutes >= s.start && this.simMinutes < s.end) : null;
    if (m.smart_room && this.state.emergencyState === 'IDLE') {
      this.state.smartRoomState = active ? 'IN_SESSION' : 'STANDBY';
    }

    // Keep the Flow Tracker page alive in demo mode by cycling the two
    // continuous-monitoring flowcharts (C and D), same as the firmware -
    // but only for modules that are actually enabled.
    if (m.environment) {
      this._flow('C', 'C_SENSOR'); this._flow('C', 'C_READ'); this._flow('C', 'C_PROCESS');
      this._flow('C', 'C_NORMAL', 'YES'); this._flow('C', 'C_DISPLAY_STATUS');
    }
    if (m.security && this.state.emergencyState === 'IDLE') {
      this._flow('D', 'D_DETECT'); this._flow('D', 'D_READ_LEVEL'); this._flow('D', 'D_COMPARE');
      this._flow('D', 'D_ABOVE_WARN', 'NO');
    }
    if (m.smart_room && this.state.smartRoomState === 'IN_SESSION') {
      this._flow('B', 'B_SENSOR_DATA'); this._flow('B', 'B_PROCESS_SENSOR'); this._flow('B', 'B_COND_NORMAL', 'YES'); this._flow('B', 'B_DISPLAY_NORMAL');
    }

    this._emitMsg({
      t: 'telemetry',
      sim_time: this._hhmm(),
      sim_scale: this.scale,
      uptime_s: Math.floor(process.uptime()),
      heap_free: 180000,
      environment: { temp_c: this.state.temp, humidity_pct: this.state.hum, motion: this.state.motion, aqi: this.state.smoke * 5 },
      security: { smoke_pct: this.state.smoke, smoke_level: this.state.smokeLevel, intrusion: false, false_alarms: 0 },
      smart_room: {
        room: ROOM, state: this.state.smartRoomState,
        subject: active ? active.subject : '', section: active ? active.section : '', faculty: active ? active.faculty : '',
        attendance: this.state.attendanceToday, abnormal: false,
      },
      waste: { fill_pct: this.state.wastePct, full: this.state.wasteFull },
      metal_detector: { threshold_pct: this.state.metalThreshold, hold_active: this.state.holdActive },
      doors: { main_entrance_unlocked: this.state.doorsUnlocked > 0, override_active: this.state.overrideActive },
      emergency: { state: this.state.emergencyState, active: this.state.overrideActive },
      wifi: { connected: this.state.wifiConnected, rssi: this.state.rssi, ip: this.state.ip },
      relay_sf03: this.state.smartRoomState !== 'STANDBY',
      attendance_today: this.state.attendanceToday,
      modules: this.modules,
    });
  }

  _handleCommand(msg) {
    const { type } = msg;
    switch (type) {
      case 'ping':
        this._ack(type, true, 'pong');
        break;
      case 'sync_users':
        this.registry = msg.users || [];
        this._ack(type, true);
        break;
      case 'sync_schedule':
        this.schedule = (msg.entries || []).filter((e) => e.room === ROOM);
        this._ack(type, true);
        break;
      case 'rfid_tap':
        this._simulateTap(msg.uid, 'web');
        this._ack(type, true);
        break;
      case 'time_scale':
        this.scale = msg.minutes_per_second || 1;
        this._ack(type, true);
        break;
      case 'jump_time':
        this.simMinutes = (msg.minutes || 0) % 1440;
        this._ack(type, true);
        break;
      case 'class_override':
        if (msg.action === 'start') {
          this.state.smartRoomState = 'IN_SESSION';
          this._flow('B', 'B_ACTIVATE'); this._flow('B', 'B_DISPLAY_STATUS'); this._flow('B', 'B_IN_SESSION');
        } else {
          this.state.smartRoomState = 'STANDBY';
          this._flow('B', 'B_SAVE_DATA'); this._flow('B', 'B_SEND_CENTRAL'); this._flow('B', 'B_STANDBY_OFF');
          this._flow('B', 'B_STANDBY_STATUS'); this._flow('B', 'B_END');
        }
        this._ack(type, true);
        break;
      case 'screening_decision':
        this.state.holdActive = false;
        this._flow('F', 'F_ALLOWED', msg.allow ? 'YES' : 'NO');
        this._flow('F', msg.allow ? 'F_ALLOW' : 'F_DENY');
        if (msg.allow) this._flow('F', 'F_END');
        this._emitMsg({
          t: 'screening', result: msg.allow ? 'allowed' : 'denied', signal_pct: 70,
          threshold_pct: this.state.metalThreshold, scans: 1, alerts: 1,
          allowed: msg.allow ? 1 : 0, denied: msg.allow ? 0 : 1, ts: this._hhmm(),
        });
        this._ack(type, true);
        break;
      case 'waste_collect':
        this.state.wastePct = 10;
        this.state.wasteFull = false;
        this._flow('H', 'H_COLLECTION'); this._flow('H', 'H_RESET'); this._flow('H', 'H_END');
        this._emitMsg({ t: 'waste', bin_id: 'lobby-main', room: 'GF-01', fill_pct: 10, state: 'collected', live: true, ts: this._hhmm() });
        this._ack(type, true);
        break;
      case 'emergency_test':
      case 'emergency_ack':
      case 'emergency_clear':
        this._handleEmergency(type);
        this._ack(type, true);
        break;
      case 'net_scenario':
        this._emitMsg({ t: 'log', msg: `Demo: network scenario '${msg.scenario}' simulated.` });
        this._ack(type, true);
        break;
      case 'net_request': {
        this._flow('E', 'E_REQUEST'); this._flow('E', 'E_AUTH'); this._flow('E', 'E_AUTHORIZED', 'YES');
        this._flow('E', 'E_CONNECT'); this._flow('E', 'E_ACCESS_LMS'); this._flow('E', 'E_EXCHANGE');
        this._flow('E', 'E_DISPLAY_INFO'); this._flow('E', 'E_END');
        this._emitMsg({ t: 'net_result', user: msg.user, role: msg.role, port: msg.port, permit: true, rule_id: 'FW-01', vlan: 'VLAN30', ts: this._hhmm() });
        this._ack(type, true);
        break;
      }
      case 'set_thresholds':
        if (msg.metal_threshold_pct) this.state.metalThreshold = msg.metal_threshold_pct;
        this._ack(type, true);
        break;
      case 'module_toggle':
        if (Object.prototype.hasOwnProperty.call(this.modules, msg.module)) {
          this.modules[msg.module] = !!msg.enabled;
          this._emitMsg({ t: 'log', msg: `Demo: module '${msg.module}' ${msg.enabled ? 'enabled' : 'disabled'}.` });
          this._ack(type, true);
        } else {
          this._ack(type, false, 'unknown module');
        }
        break;
      case 'door_override':
      case 'virtual_override':
        this._ack(type, true);
        break;
      default:
        this._ack(type, false, 'unrecognized command type (demo mode)');
    }
  }

  _handleEmergency(type) {
    if (type === 'emergency_test') {
      this.state.emergencyState = 'ACTIVE';
      this.state.overrideActive = true;
      this.state.doorsUnlocked = 1;
      this._flow('G', 'G_RECEIVED'); this._flow('G', 'G_VERIFY'); this._flow('G', 'G_CONFIRMED', 'YES');
      this._flow('G', 'G_OVERRIDE'); this._flow('G', 'G_UNLOCK'); this._flow('G', 'G_ALARM');
      this._flow('G', 'G_DISPLAY_ALERT'); this._flow('G', 'G_NOTIFY'); this._flow('G', 'G_ADMIN_RECEIVE');
      this._emitMsg({ t: 'alert', module: 'G', severity: 'critical', phase: 'active', state: 'ACTIVE', reason: 'Emergency drill (dashboard test)', drill: true, ts: this._hhmm() });
    } else if (type === 'emergency_ack') {
      if (this.state.emergencyState === 'ACTIVE') {
        this.state.emergencyState = 'RESPONSE';
        this._flow('G', 'G_RESPONSE');
        this._emitMsg({ t: 'alert', module: 'G', severity: 'critical', phase: 'acknowledged', state: 'RESPONSE', reason: '', drill: false, ts: this._hhmm() });
      }
    } else if (type === 'emergency_clear') {
      if (this.state.emergencyState === 'RESPONSE') {
        this.state.emergencyState = 'IDLE';
        this.state.overrideActive = false;
        this.state.doorsUnlocked = 0;
        this._flow('G', 'G_CLEARED', 'YES'); this._flow('G', 'G_RESET'); this._flow('G', 'G_RETURN_DOORS'); this._flow('G', 'G_END');
        this._emitMsg({ t: 'alert', module: 'G', severity: 'info', phase: 'cleared', state: 'IDLE', reason: '', drill: false, ts: this._hhmm() });
      }
    }
  }

  _simulateTap(uid, source) {
    this._flow('A', 'A_START'); this._flow('A', 'A_READER'); this._flow('A', 'A_PROCESS');
    const user = this.registry.find((u) => u.uid === uid);
    if (!user) {
      this._flow('A', 'A_CHECK', 'NO'); this._flow('A', 'A_INVALID');
      this._emitMsg({ t: 'rfid', uid, source, result: 'denied', name: '', role: '', room: '', ts: this._hhmm() });
      return;
    }
    this._flow('A', 'A_CHECK', 'YES'); this._flow('A', 'A_RECORD'); this._flow('A', 'A_DATA');
    this._flow('A', 'A_DB'); this._flow('A', 'A_REGISTRAR'); this._flow('A', 'A_END');
    this.state.attendanceToday++;
    this._emitMsg({ t: 'rfid', uid, source, result: 'granted', name: user.name, role: user.role, room: user.room, ts: this._hhmm() });
    this._emitMsg({ t: 'attendance', uid, name: user.name, role: user.role, room: user.room || 'Main Entrance', ts: this._hhmm(), source });
  }

  _ack(cmd, ok, detail = '') {
    this._emitMsg({ t: 'ack', cmd, ok, detail });
  }
}

File: server/demo_device.py
"""Smart Learning Center - demo-mode device generator.

Mimics the ESP32's wire protocol (same 'connected' / 'disconnected' /
'message' / 'raw' events and async send(line) method as bridge.py) so the
rest of main.py does not need to know whether it is talking to the real
Wokwi simulation or this generator. Used when BRIDGE_MODE=demo, so the
dashboard can be rehearsed without Wokwi running. The UI must show a
persistent DEMO MODE banner whenever this is active - see main.py.
"""
import asyncio
import json
import random
import time

from emitter import AsyncEmitter

ROOM = "SF-03"


class DemoDevice(AsyncEmitter):
    def __init__(self):
        super().__init__()
        self.connected = False
        self.sim_minutes = 7 * 60 + 10
        self.scale = 1
        self.registry = []
        self.schedule = []
        self._task = None
        self._start_time = time.monotonic()

        self.state = {
            "temp": 24.0, "hum": 55.0, "motion": False,
            "smoke": 6.0, "smoke_level": 0,
            "waste_pct": 15.0, "waste_full": False,
            "metal_threshold": 55.0, "hold_active": False,
            "doors_unlocked": 0, "override_active": False,
            "emergency_state": "IDLE",
            "smart_room_state": "STANDBY",
            "attendance_today": 0,
            "wifi_connected": True, "rssi": -58, "ip": "10.0.0.42",
        }

        self.modules = {
            "rfid": True, "smart_room": True, "environment": True, "security": True,
            "metal_detector": True, "waste": True, "network": True,
        }

    def start(self):
        self.connected = True
        self._task = asyncio.create_task(self._run())

    def stop(self):
        self.connected = False
        if self._task:
            self._task.cancel()

    def is_connected(self) -> bool:
        return self.connected

    async def send(self, line: str) -> bool:
        if not line.startswith("CMD "):
            return False
        try:
            msg = json.loads(line[4:])
        except json.JSONDecodeError:
            return False
        await self._handle_command(msg)
        return True

    async def _run(self):
        await asyncio.sleep(0.05)
        await self.emit("connected")
        await self._emit_msg({"t": "boot", "device": "DEMO-ESP32", "fw": "demo-1.0.0", "modules": "A,B,C,D,E,F,G,H"})
        await self._emit_msg({"t": "log", "msg": "Demo device generator started (BRIDGE_MODE=demo). No Wokwi simulation is running."})
        while True:
            await asyncio.sleep(1)
            await self._tick()

    async def _emit_msg(self, obj):
        await self.emit("message", obj)

    async def _flow(self, m, s, v=None):
        msg = {"t": "flow", "m": m, "s": s}
        if v:
            msg["v"] = v
        await self._emit_msg(msg)

    def _hhmm(self) -> str:
        h = (self.sim_minutes // 60) % 24
        m = self.sim_minutes % 60
        return f"{h:02d}:{m:02d}"

    async def _tick(self):
        self.sim_minutes = (self.sim_minutes + self.scale) % 1440
        m = self.modules
        s = self.state

        if m["environment"]:
            s["temp"] = max(18.0, min(32.0, s["temp"] + (random.random() - 0.5) * 0.3))
            s["hum"] = max(30.0, min(85.0, s["hum"] + (random.random() - 0.5) * 0.6))
            s["motion"] = random.random() < 0.3
        if m["security"]:
            s["smoke"] = max(2.0, min(18.0, s["smoke"] + (random.random() - 0.5) * 2))
        if m["waste"] and not s["waste_full"]:
            s["waste_pct"] = min(100.0, s["waste_pct"] + 0.04)
            s["waste_full"] = s["waste_pct"] >= 80

        active = None
        if m["smart_room"]:
            active = next((e for e in self.schedule if self.sim_minutes >= e["start"] and self.sim_minutes < e["end"]), None)
            if s["emergency_state"] == "IDLE":
                s["smart_room_state"] = "IN_SESSION" if active else "STANDBY"

        # Keep dashboards showing live flow activity, same as the firmware -
        # but only for modules that are actually enabled.
        if m["environment"]:
            for node in ("C_SENSOR", "C_READ", "C_PROCESS"):
                await self._flow("C", node)
            await self._flow("C", "C_NORMAL", "YES")
            await self._flow("C", "C_DISPLAY_STATUS")
        if m["security"] and s["emergency_state"] == "IDLE":
            for node in ("D_DETECT", "D_READ_LEVEL", "D_COMPARE"):
                await self._flow("D", node)
            await self._flow("D", "D_ABOVE_WARN", "NO")
        if m["smart_room"] and s["smart_room_state"] == "IN_SESSION":
            for node in ("B_SENSOR_DATA", "B_PROCESS_SENSOR"):
                await self._flow("B", node)
            await self._flow("B", "B_COND_NORMAL", "YES")
            await self._flow("B", "B_DISPLAY_NORMAL")

        await self._emit_msg({
            "t": "telemetry",
            "sim_time": self._hhmm(),
            "sim_scale": self.scale,
            "uptime_s": int(time.monotonic() - self._start_time),
            "heap_free": 180000,
            "environment": {"temp_c": s["temp"], "humidity_pct": s["hum"], "motion": s["motion"], "aqi": s["smoke"] * 5},
            "security": {"smoke_pct": s["smoke"], "smoke_level": s["smoke_level"], "intrusion": False, "false_alarms": 0},
            "smart_room": {
                "room": ROOM, "state": s["smart_room_state"],
                "subject": active["subject"] if active else "", "section": active["section"] if active else "",
                "faculty": active["faculty"] if active else "",
                "attendance": s["attendance_today"], "abnormal": False,
            },
            "waste": {"fill_pct": s["waste_pct"], "full": s["waste_full"]},
            "metal_detector": {"threshold_pct": s["metal_threshold"], "hold_active": s["hold_active"]},
            "doors": {"main_entrance_unlocked": s["doors_unlocked"] > 0, "override_active": s["override_active"]},
            "emergency": {
                "state": s["emergency_state"],
                "active": s["emergency_state"] == "ACTIVE",
                "pending": s["emergency_state"] == "PENDING",
            },
            "wifi": {"connected": s["wifi_connected"], "rssi": s["rssi"], "ip": s["ip"]},
            "relay_sf03": s["smart_room_state"] != "STANDBY",
            "attendance_today": s["attendance_today"],
            "modules": self.modules,
        })

    async def _handle_command(self, msg):
        type_ = msg.get("type")
        s = self.state

        if type_ == "ping":
            await self._ack(type_, True, "pong")
        elif type_ == "sync_users":
            self.registry = msg.get("users") or []
            await self._ack(type_, True)
        elif type_ == "sync_schedule":
            self.schedule = [e for e in (msg.get("entries") or []) if e.get("room") == ROOM]
            await self._ack(type_, True)
        elif type_ == "rfid_tap":
            await self._simulate_tap(msg.get("uid", ""), "web")
            await self._ack(type_, True)
        elif type_ == "time_scale":
            self.scale = msg.get("minutes_per_second") or 1
            await self._ack(type_, True)
        elif type_ == "jump_time":
            self.sim_minutes = (msg.get("minutes") or 0) % 1440
            await self._ack(type_, True)
        elif type_ == "class_override":
            if msg.get("action") == "start":
                s["smart_room_state"] = "IN_SESSION"
                for node in ("B_ACTIVATE", "B_DISPLAY_STATUS", "B_IN_SESSION"):
                    await self._flow("B", node)
            else:
                s["smart_room_state"] = "STANDBY"
                for node in ("B_SAVE_DATA", "B_SEND_CENTRAL", "B_STANDBY_OFF", "B_STANDBY_STATUS", "B_END"):
                    await self._flow("B", node)
            await self._ack(type_, True)
        elif type_ == "screening_decision":
            allow = bool(msg.get("allow"))
            s["hold_active"] = False
            await self._flow("F", "F_ALLOWED", "YES" if allow else "NO")
            await self._flow("F", "F_ALLOW" if allow else "F_DENY")
            if allow:
                await self._flow("F", "F_END")
            await self._emit_msg({
                "t": "screening", "result": "allowed" if allow else "denied", "signal_pct": 70,
                "threshold_pct": s["metal_threshold"], "scans": 1, "alerts": 1,
                "allowed": 1 if allow else 0, "denied": 0 if allow else 1, "ts": self._hhmm(),
            })
            await self._ack(type_, True)
        elif type_ == "waste_collect":
            s["waste_pct"] = 10.0
            s["waste_full"] = False
            for node in ("H_COLLECTION", "H_RESET", "H_END"):
                await self._flow("H", node)
            await self._emit_msg({"t": "waste", "bin_id": "lobby-main", "room": "GF-01", "fill_pct": 10, "state": "collected", "live": True, "ts": self._hhmm()})
            await self._ack(type_, True)
        elif type_ in ("emergency_test", "emergency_confirm", "emergency_dismiss", "emergency_clear"):
            await self._handle_emergency(type_)
            await self._ack(type_, True)
        elif type_ == "net_scenario":
            await self._emit_msg({"t": "log", "msg": f"Demo: network scenario '{msg.get('scenario')}' simulated."})
            await self._ack(type_, True)
        elif type_ == "net_request":
            for node in ("E_REQUEST", "E_AUTH"):
                await self._flow("E", node)
            await self._flow("E", "E_AUTHORIZED", "YES")
            for node in ("E_CONNECT", "E_ACCESS_LMS", "E_EXCHANGE", "E_DISPLAY_INFO", "E_END"):
                await self._flow("E", node)
            await self._emit_msg({"t": "net_result", "user": msg.get("user"), "role": msg.get("role"), "port": msg.get("port"), "permit": True, "rule_id": "FW-01", "vlan": "VLAN30", "ts": self._hhmm()})
            await self._ack(type_, True)
        elif type_ == "set_thresholds":
            if msg.get("metal_threshold_pct"):
                s["metal_threshold"] = msg["metal_threshold_pct"]
            await self._ack(type_, True)
        elif type_ == "module_toggle":
            name = msg.get("module")
            if name in self.modules:
                self.modules[name] = bool(msg.get("enabled"))
                await self._emit_msg({"t": "log", "msg": f"Demo: module '{name}' {'enabled' if msg.get('enabled') else 'disabled'}."})
                await self._ack(type_, True)
            else:
                await self._ack(type_, False, "unknown module")
        elif type_ in ("door_override", "virtual_override"):
            await self._ack(type_, True)
        else:
            await self._ack(type_, False, "unrecognized command type (demo mode)")

    async def _handle_emergency(self, type_):
        s = self.state
        if type_ == "emergency_test":
            if s["emergency_state"] != "IDLE":
                return
            s["emergency_state"] = "ACTIVE"
            s["override_active"] = True
            s["doors_unlocked"] = 1
            for node in ("G_RECEIVED", "G_VERIFY"):
                await self._flow("G", node)
            await self._flow("G", "G_CONFIRMED", "YES")
            for node in ("G_OVERRIDE", "G_UNLOCK", "G_ALARM", "G_DISPLAY_ALERT", "G_NOTIFY", "G_ADMIN_RECEIVE"):
                await self._flow("G", node)
            await self._emit_msg({"t": "alert", "module": "G", "severity": "critical", "phase": "active", "state": "ACTIVE", "reason": "Emergency drill (dashboard test)", "drill": True, "ts": self._hhmm()})
        elif type_ == "emergency_confirm":
            if s["emergency_state"] != "PENDING":
                return
            s["emergency_state"] = "ACTIVE"
            s["override_active"] = True
            s["doors_unlocked"] = 1
            for node in ("G_OVERRIDE", "G_UNLOCK", "G_ALARM", "G_DISPLAY_ALERT", "G_NOTIFY", "G_ADMIN_RECEIVE"):
                await self._flow("G", node)
            await self._emit_msg({"t": "alert", "module": "G", "severity": "critical", "phase": "active", "state": "ACTIVE", "reason": "", "drill": False, "ts": self._hhmm()})
        elif type_ == "emergency_dismiss":
            if s["emergency_state"] != "PENDING":
                return
            s["emergency_state"] = "IDLE"
            await self._flow("G", "G_LOG_CANCEL")
            await self._flow("G", "G_NORMAL_STATUS")
            await self._emit_msg({"t": "alert", "module": "G", "severity": "info", "phase": "dismissed", "state": "IDLE", "reason": "", "drill": False, "ts": self._hhmm()})
        elif type_ == "emergency_clear":
            if s["emergency_state"] != "ACTIVE":
                return
            s["emergency_state"] = "IDLE"
            s["override_active"] = False
            s["doors_unlocked"] = 0
            for node in ("G_RESET", "G_RETURN_DOORS", "G_END"):
                await self._flow("G", node)
            await self._emit_msg({"t": "alert", "module": "G", "severity": "info", "phase": "cleared", "state": "IDLE", "reason": "", "drill": False, "ts": self._hhmm()})

    async def _simulate_tap(self, uid, source):
        for node in ("A_START", "A_READER", "A_PROCESS"):
            await self._flow("A", node)
        user = next((u for u in self.registry if u.get("uid") == uid), None)
        if not user:
            await self._flow("A", "A_CHECK", "NO")
            await self._flow("A", "A_INVALID")
            await self._emit_msg({"t": "rfid", "uid": uid, "source": source, "result": "denied", "name": "", "role": "", "room": "", "ts": self._hhmm()})
            return
        await self._flow("A", "A_CHECK", "YES")
        for node in ("A_RECORD", "A_DATA", "A_DB", "A_REGISTRAR", "A_END"):
            await self._flow("A", node)
        self.state["attendance_today"] += 1
        await self._emit_msg({"t": "rfid", "uid": uid, "source": source, "result": "granted", "name": user["name"], "role": user["role"], "room": user.get("room", ""), "ts": self._hhmm()})
        await self._emit_msg({"t": "attendance", "uid": uid, "name": user["name"], "role": user["role"], "room": user.get("room") or "Main Entrance", "ts": self._hhmm(), "source": source})

    async def _ack(self, cmd, ok, detail=""):
        await self._emit_msg({"t": "ack", "cmd": cmd, "ok": ok, "detail": detail})
